contracts/common/SafeERC20.sol

```diff
@@ -0,0 +1,68 @@
+// Inspired by AdEx (https://github.com/AdExNetwork/adex-protocol-eth/blob/b9df617829661a7518ee10f4cb6c4108659dd6d5/contracts/libs/SafeERC20.sol)
+// and 0x (https://github.com/0xProject/0x-monorepo/blob/737d1dc54d72872e24abce5a1dbe1b66d35fa21a/contracts/protocol/contracts/protocol/AssetProxy/ERC20Proxy.sol#L143)
+
+pragma solidity ^0.4.24;
+
+import "../lib/token/ERC20.sol";
+
+// Same as ERC20 except the return values of these interfaces have been removed to allow us to
+// manually check them
+interface GeneralERC20 {
+    function transfer(address to, uint256 value) external;
+    function transferFrom(address from, address to, uint256 value) external;
+    function approve(address spender, uint256 value) external;
+}
+
+
+library SafeERC20 {
+    function checkSuccess() private pure returns (bool ret) {
+        assembly {
+            // Check number of bytes returned from last function call
+            switch returndatasize
+
+            // No bytes returned: assume success
+            case 0x0 {
+                ret := 1
+            }
+
+            // 32 bytes returned: check if non-zero
+            case 0x20 {
+                // Copy 32 bytes into scratch space
+                returndatacopy(0x0, 0x0, 0x20)
```

Comment:
I like how by doing this optimization we save a few `mload` and `mstore`. Given that we read this memory just immediately after we should be okey using this memory address, but I always feel a little uneasy that something else may be counting on this memory space not changing.

Reply:
> Something else may be counting on this memory space not changing

I really hope not... [`0x00 - 0x3f` is scratch space for doing keccaks](https://solidity.readthedocs.io/en/v0.4.24/miscellaneous.html#layout-in-memory), so you should never rely on storing something there unless you know you're going to retrieve it immediately.

Reply:
Actually it looks like the scratch space could be used by solidity even in-between inline assembly statements, so it's not safe to ever use (except for throwaway values). Will use the free memory pointer.